chain/client/src/test_utils.rs

```diff
@@ -1102,7 +1102,151 @@ pub struct TestEnv {
+    ///
+    /// The vector must have the same number of elements as they are clients.
+    /// If that does not hold, `build` method will panic.
+    pub fn network_adapters(mut self, adapters: Vec<Arc<MockNetworkAdapter>>) -> Self {
+        self.network_adapters = Some(adapters);
+        self
+    }
+
+    /// Constructs new `TestEnv` structure.
+    ///
+    /// If no clients were configured (either through count or vector) one
+    /// client is created.  Similarly, if no validator seats were configured,
+    /// one seat is configured.
+    ///
+    /// Panics if `runtime_adapters` or `network_adapters` methods were used and
+    /// the length of the vectors passed to them did not equal number of
+    /// configured clients.
+    pub fn build(self) -> TestEnv {
+        let chain_genesis = self.chain_genesis;
+        let clients = self.clients;
+        let num_clients = clients.len();
+        let validators = self.validators;
+        let num_validators = validators.len();
+        let network_adapters = self
+            .network_adapters
+            .unwrap_or_else(|| (0..num_clients).map(|_| Arc::new(Default::default())).collect());
+        assert!(clients.len() == network_adapters.len());
+        let clients = match self.runtime_adapters {
+            None => clients
+                .into_iter()
+                .zip(network_adapters.iter())
+                .map(|(account_id, network_adapter)| {
+                    setup_client(
+                        create_test_store(),
+                        vec![validators.clone()],
+                        1,
+                        1,
+                        Some(account_id),
+                        false,
+                        network_adapter.clone(),
+                        chain_genesis.clone(),
+                    )
+                })
+                .collect(),
+            Some(runtime_adapters) => {
+                assert!(clients.len() == runtime_adapters.len());
+                clients
+                    .into_iter()
+                    .zip((&network_adapters).iter())
+                    .zip(runtime_adapters.into_iter())
+                    .map(|((account_id, network_adapter), runtime_adapter)| {
+                        setup_client_with_runtime(
+                            u64::try_from(num_validators).unwrap(),
+                            Some(account_id),
+                            false,
+                            network_adapter.clone(),
+                            chain_genesis.clone(),
+                            runtime_adapter,
+                        )
+                    })
+                    .collect()
+            }
+        };
+
+        TestEnv { chain_genesis, validators, network_adapters, clients }
+    }
+
+    fn default_formatter(id: usize) -> std::string::String {
+        format!("test{}", id)
+    }
+
+    fn make_accounts<F>(count: usize, formatter: F) -> Vec<AccountId>
```

Comment:
We always use `default_formatter`, so I'd just YAGNI this. 

If the user wants to customize it, i'd let them provide the actual names, rather than a formatter.

Reply:
In early version of this code there was a `clients_with_formatter` function which took number of clients and a formatter function. While it’s true that with `clients` method where the vector can be provided directly support for custom formatters is unnecessary but it seems that the most common case will be to to simply use `clients_count(number)` and the second most common will be to use `clients_with_formatter(number, |id| format("blah{}", id)`. Perhaps it’s `clients` method that we’re not gonna need?

Reply:
> the second most common will be to use clients_with_formatter(number, |id| format("blah{}", id)

Why would this be common? It seems that there shouldn't be a need to customize the names, outside of the special cases like #4771, where overriding the whole thing should be enough. 

I feel `clients` is needed because it's the most general -- even if we call it zero times, it's important that we can customize this behavior when we suddenly need this for some future investigation. This avoids [midlayer mistake](https://lwn.net/Articles/336262/) -- providing convenience API without exposing the full capabilities of the underlying platform. 

Playing around this ideas, I got the following: 

```diff
diff --git a/chain/client/src/test_utils.rs b/chain/client/src/test_utils.rs
index 723584675..9d97d3ece 100644
--- a/chain/client/src/test_utils.rs
+++ b/chain/client/src/test_utils.rs
@@ -1117,11 +1117,13 @@ impl TestEnvBuilder {
     fn new(chain_genesis: ChainGenesis) -> Self {
         Self {
             chain_genesis,
-            clients: Self::make_accounts(1, Self::default_formatter),
-            validators: Self::make_accounts(1, Self::default_formatter),
+            clients: Vec::new(),
+            validators: Vec::new(),
             runtime_adapters: None,
             network_adapters: None,
         }
+        .clients_count(1)
+        .validator_seats(1)
     }
 
     /// Sets list of client `AccountId`s to the one provided.  Panics if the
@@ -1132,10 +1134,11 @@ impl TestEnvBuilder {
         self
     }
 
-    /// Sets number of clients to given one.  Each client will use `AccountId`
-    /// in the form `test{}` where `{}` will count from zero.
+    /// Sets number of clients to given one.
     pub fn clients_count(self, num: usize) -> Self {
-        self.clients(Self::make_accounts(num, Self::default_formatter))
+        let clients =
+            (0..num).map(|i| AccountId::try_from(format!("test{}", i)).unwrap()).collect();
+        self.clients(clients)
     }
 
     /// Sets list of validator `AccountId`s to the one provided.  Panics if the
@@ -1146,10 +1149,10 @@ impl TestEnvBuilder {
         self
     }
 
-    /// Sets number of validator seats to given one.  Each validator will use
-    /// `AccountId` in the form `test{}` where `{}` will count from zero.
+    /// Sets the first `num` accounts as validators.
     pub fn validator_seats(self, num: usize) -> Self {
-        self.validators(Self::make_accounts(num, Self::default_formatter))
+        let validators = self.clients[..num].to_vec();
+        self.validators(validators)
     }
 
     /// Specifies custom runtime adaptors for each client.  This allows us to
@@ -1229,17 +1232,6 @@ impl TestEnvBuilder {
 
         TestEnv { chain_genesis, validators, network_adapters, clients }
     }
-
-    fn default_formatter(id: usize) -> std::string::String {
-        format!("test{}", id)
-    }
-
-    fn make_accounts<F>(count: usize, formatter: F) -> Vec<AccountId>
-    where
-        F: Fn(usize) -> std::string::String,
-    {
-        (0..count).map(|i| AccountId::try_from(formatter(i)).unwrap()).collect()
-    }
 }
 
 impl TestEnv {

```

I've almo made a couple observations while playing with it:

* there's slight naming inconsistensy where we use count, num, and seats
* not sure if we should document the exact format of test account naming. Seems better to provide accessor to resulting `AccountId` structs. Reliance on specific names is exactly what caused the original issues with different names not working.

Reply:
You’re not wrong and my assertion isn’t really supported by hard data. On the other hand, we have no examples of tests where we use account ids which don’t follow simple pattern while we have an example of test which wants to change the pattern. ;)

I’m also not exactly convinced that we should be worried about the midlayer mistake. We can have both `clients` and `clients_with_formatter`. While that ‘bloats’ the builder, I think it’s acceptable for a helper test code whose whole purpose is to provide convenience.

> there's slight naming inconsistensy where we use count, num, and seats

I was going back and forth between `validator_seats` and `validators_count`. I don’t really have any good reasons for either to be honest. I first went with count but then saw bunch of code use seats. Don’t really have an opinion.

> Seems better to provide accessor to resulting AccountId structs.

Yes, I've been looking into that as well though I think such accessor belongs to `TestEnv` rather than the builder.

PS. Note that there are tests where number of validator seats is greater than number of clients, see `test-utils/state-viewer/src/state_dump.rs` so `validators = self.clients[..num].to_vec();` is sadly not valid.